### PR TITLE
efi: use 4k pages on aarch64

### LINKED ISSUE #23789: /usr/bin/ld.bfd: warning: src/boot/efi/systemd-bootaa64.elf has a LOAD segment with RWX permissions
Building systemd for arm64 with a recent version of binutils (2.38.50.20220615-4ubuntu1) fails as a result of the following warning:

```
[67/2388] /usr/bin/cc -o src/boot/efi/systemd-bootaa64.elf -Wno-format-signedness -Wno-missing-field-initializers -Wno-unused-parameter -Wdate-time -Wendif-labels -Werror=format=2 -Werror=implicit-function-declaration -Werror=incompatible-pointer-types -Werror=int-conversion -Werror=overflow -Werror=override-init -Werror=return-type -Werror=shift-count-overflow -Werror=shift-overflow=2 -Werror=undef -Wfloat-equal -Wimplicit-fallthrough=5 -Winit-self -Wlogical-op -Wmissing-include-dirs -Wmissing-noreturn -Wnested-externs -Wold-style-definition -Wpointer-arith -Wredundant-decls -Wshadow -Wstrict-aliasing=2 -Wstrict-prototypes -Wsuggest-attribute=noreturn -Wunused-function -Wwrite-strings -Wno-unused-result -fno-stack-protector -fno-strict-aliasing -fpic -fwide-exec-charset=UCS2 -Wall -Wextra -Wsign-compare -nostdlib -std=gnu11 -ffreestanding -fshort-wchar -fvisibility=hidden -isystem /usr/include/efi -isystem /usr/include/efi/aarch64 -I /<<PKGBUILDDIR>>/src/fundamental -DSD_BOOT -DGNU_EFI_USE_MS_ABI -include src/boot/efi/efi_config.h -include version.h -flto -O2 -ffile-prefix-map=/<<PKGBUILDDIR>>=. -flto=auto -flto=auto -fuse-ld=bfd -L /usr/lib -nostdlib -T /usr/lib/elf_aarch64_efi.lds -Wl,--build-id=sha1 -Wl,--fatal-warnings -Wl,--no-undefined -Wl,--warn-common -Wl,-Bsymbolic -z nocombreloc /usr/lib/crt0-efi-aarch64.o -shared -Wl,--defsym=EFI_SUBSYSTEM=0xa src/boot/efi/bootspec-fundamental.c.o src/boot/efi/efivars-fundamental.c.o src/boot/efi/sha256.c.o src/boot/efi/string-util-fundamental.c.o src/boot/efi/assert.c.o src/boot/efi/devicetree.c.o src/boot/efi/disk.c.o src/boot/efi/graphics.c.o src/boot/efi/measure.c.o src/boot/efi/pe.c.o src/boot/efi/secure-boot.c.o src/boot/efi/ticks.c.o src/boot/efi/util.c.o src/boot/efi/boot.c.o src/boot/efi/console.c.o src/boot/efi/drivers.c.o src/boot/efi/random-seed.c.o src/boot/efi/shim.c.o src/boot/efi/xbootldr.c.o src/boot/efi/bcd.c.o -lefi -lgnuefi -lgcc
FAILED: src/boot/efi/systemd-bootaa64.elf 
/usr/bin/cc -o src/boot/efi/systemd-bootaa64.elf -Wno-format-signedness -Wno-missing-field-initializers -Wno-unused-parameter -Wdate-time -Wendif-labels -Werror=format=2 -Werror=implicit-function-declaration -Werror=incompatible-pointer-types -Werror=int-conversion -Werror=overflow -Werror=override-init -Werror=return-type -Werror=shift-count-overflow -Werror=shift-overflow=2 -Werror=undef -Wfloat-equal -Wimplicit-fallthrough=5 -Winit-self -Wlogical-op -Wmissing-include-dirs -Wmissing-noreturn -Wnested-externs -Wold-style-definition -Wpointer-arith -Wredundant-decls -Wshadow -Wstrict-aliasing=2 -Wstrict-prototypes -Wsuggest-attribute=noreturn -Wunused-function -Wwrite-strings -Wno-unused-result -fno-stack-protector -fno-strict-aliasing -fpic -fwide-exec-charset=UCS2 -Wall -Wextra -Wsign-compare -nostdlib -std=gnu11 -ffreestanding -fshort-wchar -fvisibility=hidden -isystem /usr/include/efi -isystem /usr/include/efi/aarch64 -I /<<PKGBUILDDIR>>/src/fundamental -DSD_BOOT -DGNU_EFI_USE_MS_ABI -include src/boot/efi/efi_config.h -include version.h -flto -O2 -ffile-prefix-map=/<<PKGBUILDDIR>>=. -flto=auto -flto=auto -fuse-ld=bfd -L /usr/lib -nostdlib -T /usr/lib/elf_aarch64_efi.lds -Wl,--build-id=sha1 -Wl,--fatal-warnings -Wl,--no-undefined -Wl,--warn-common -Wl,-Bsymbolic -z nocombreloc /usr/lib/crt0-efi-aarch64.o -shared -Wl,--defsym=EFI_SUBSYSTEM=0xa src/boot/efi/bootspec-fundamental.c.o src/boot/efi/efivars-fundamental.c.o src/boot/efi/sha256.c.o src/boot/efi/string-util-fundamental.c.o src/boot/efi/assert.c.o src/boot/efi/devicetree.c.o src/boot/efi/disk.c.o src/boot/efi/graphics.c.o src/boot/efi/measure.c.o src/boot/efi/pe.c.o src/boot/efi/secure-boot.c.o src/boot/efi/ticks.c.o src/boot/efi/util.c.o src/boot/efi/boot.c.o src/boot/efi/console.c.o src/boot/efi/drivers.c.o src/boot/efi/random-seed.c.o src/boot/efi/shim.c.o src/boot/efi/xbootldr.c.o src/boot/efi/bcd.c.o -lefi -lgnuefi -lgcc
/usr/bin/ld.bfd: warning: src/boot/efi/systemd-bootaa64.elf has a LOAD segment with RWX permissions
collect2: error: ld returned 1 exit status
```

This is a new linker warning, added in https://sourceware.org/git/?p=binutils-gdb.git;a=commit;h=ba951afb99912da01a6e8434126b8fac7aa75107.

### PR DESCRIPTION
**Pull request by keszybz**
Fixes #23789.

This patch is only complile-tested by me. I don't have the appropriate hardware.

### CODE REVIEW

**Comment by keszybz** [comment]
@enr0n, @ozbenh, @FantasqueX, please check if this works for you.

**Comment by medhefgo** [comment]
The way my math works out is that anything that is 64k aligned is also 4k aligned, so how is this supposed to work?

I say this is a gnu-efi issue as they don't properly align ELF segments in the linker script. Afaik, ELF segments get ignored by the PE conversion, so the whole warning in itself is useless for us, which is why we disable this warning if supported. So I am more interested in why this warning is still happening?

**Comment by keszybz** [comment]
You are of course correct that 64k is a specific case of 4k alignment. But there seems to be an effect other than the offsets used. In particular, the alignment is saved as metadata. This might be enough to generate the warning:
```console
$ diffoscope --diff-context=1 build/src/boot/efi/linuxaa64.elf.stub build/src/boot/efi/linuxaa64.elf.stub.4096
--- build/src/boot/efi/linuxaa64.elf.stub
+++ build/src/boot/efi/linuxaa64.elf.stub.4096
│┄ File has been modified after NT_GNU_BUILD_ID has been applied.
├── readelf --wide --file-header {}
│ @@ -12,3 +12,3 @@
│    Start of program headers:          64 (bytes into file)
│ -  Start of section headers:          426104 (bytes into file)
│ +  Start of section headers:          364664 (bytes into file)
│    Flags:                             0x0
├── readelf --wide --program-header {}
│ @@ -7,5 +7,5 @@
│    Type           Offset   VirtAddr           PhysAddr           FileSiz  MemSiz   Flg Align
│ -  LOAD           0x010000 0x0000000000000000 0x0000000000000000 0x01d024 0x01d024 RWE 0x10000
│ -  DYNAMIC        0x029b30 0x0000000000019b30 0x0000000000019b30 0x000100 0x000100 RW  0x8
│ -  NOTE           0x02d000 0x000000000001d000 0x000000000001d000 0x000024 0x000024 R   0x4
│ +  LOAD           0x001000 0x0000000000000000 0x0000000000000000 0x01d024 0x01d024 RWE 0x1000
│ +  DYNAMIC        0x01ab30 0x0000000000019b30 0x0000000000019b30 0x000100 0x000100 RW  0x8
│ +  NOTE           0x01e000 0x000000000001d000 0x000000000001d000 0x000024 0x000024 R   0x4
│    GNU_STACK      0x000000 0x0000000000000000 0x0000000000000000 0x000000 0x000000 RW  0x10
├── readelf --wide --sections {}
│ @@ -1,2 +1,2 @@
│ -There are 23 section headers, starting at offset 0x68078:
│ +There are 23 section headers, starting at offset 0x59078:
│  
│ @@ -5,24 +5,24 @@
│    [ 0]                   NULL            0000000000000000 000000 000000 00      0   0  0
│ -  [ 1] .text             PROGBITS        0000000000000000 010000 019b30 00  AX  0   0 4096
│ -  [ 2] .dynamic          DYNAMIC         0000000000019b30 029b30 000100 10  WA  7   0  8
│ -  [ 3] .data             PROGBITS        000000000001a000 02a000 0003f0 08  WA  0   0 4096
│ -  [ 4] .rela.got         RELA            000000000001a3f0 02a3f0 000090 18   A  6   0  8
│ -  [ 5] .rela.data        RELA            000000000001a480 02a480 000690 18  AI  6   3  8
│ -  [ 6] .dynsym           DYNSYM          000000000001b000 02b000 000210 18   A  7   3  8
│ -  [ 7] .dynstr           STRTAB          000000000001c000 02c000 0000cf 00   A  0   0  1
│ -  [ 8] .gnu.hash         GNU_HASH        000000000001c0d0 02c0d0 0000b0 00   A  6   0  8
│ -  [ 9] .sbat             PROGBITS        000000000001c180 02c180 0000e5 00   A  0   0  8
│ -  [10] .sdmagic          PROGBITS        000000000001c268 02c268 000037 00   A  0   0  8
│ -  [11] .note.gnu.build-id NOTE            000000000001d000 02d000 000024 00   A  0   0  4
│ -  [12] .comment          PROGBITS        0000000000000000 02d024 00005c 01  MS  0   0  1
│ -  [13] .debug_info       PROGBITS        0000000000000000 02d080 026c62 00      0   0  1
│ -  [14] .debug_abbrev     PROGBITS        0000000000000000 053ce2 003d63 00      0   0  1
│ -  [15] .debug_aranges    PROGBITS        0000000000000000 057a45 0004a0 00      0   0  1
│ -  [16] .debug_line       PROGBITS        0000000000000000 057ee5 006a1e 00      0   0  1
│ -  [17] .debug_str        PROGBITS        0000000000000000 05e903 004950 01  MS  0   0  1
│ -  [18] .debug_line_str   PROGBITS        0000000000000000 063253 000596 01  MS  0   0  1
│ -  [19] .debug_rnglists   PROGBITS        0000000000000000 0637e9 0000ed 00      0   0  1
│ -  [20] .symtab           SYMTAB          0000000000000000 0638d8 0035e8 18     21 556  8
│ -  [21] .strtab           STRTAB          0000000000000000 066ec0 0010d4 00      0   0  1
│ -  [22] .shstrtab         STRTAB          0000000000000000 067f94 0000e4 00      0   0  1
│ +  [ 1] .text             PROGBITS        0000000000000000 001000 019b30 00  AX  0   0 4096
│ +  [ 2] .dynamic          DYNAMIC         0000000000019b30 01ab30 000100 10  WA  7   0  8
│ +  [ 3] .data             PROGBITS        000000000001a000 01b000 0003f0 08  WA  0   0 4096
│ +  [ 4] .rela.got         RELA            000000000001a3f0 01b3f0 000090 18   A  6   0  8
│ +  [ 5] .rela.data        RELA            000000000001a480 01b480 000690 18  AI  6   3  8
│ +  [ 6] .dynsym           DYNSYM          000000000001b000 01c000 000210 18   A  7   3  8
│ +  [ 7] .dynstr           STRTAB          000000000001c000 01d000 0000cf 00   A  0   0  1
│ +  [ 8] .gnu.hash         GNU_HASH        000000000001c0d0 01d0d0 0000b0 00   A  6   0  8
│ +  [ 9] .sbat             PROGBITS        000000000001c180 01d180 0000e5 00   A  0   0  8
│ +  [10] .sdmagic          PROGBITS        000000000001c268 01d268 000037 00   A  0   0  8
│ +  [11] .note.gnu.build-id NOTE            000000000001d000 01e000 000024 00   A  0   0  4
│ +  [12] .comment          PROGBITS        0000000000000000 01e024 00005c 01  MS  0   0  1
│ +  [13] .debug_info       PROGBITS        0000000000000000 01e080 026c62 00      0   0  1
│ +  [14] .debug_abbrev     PROGBITS        0000000000000000 044ce2 003d63 00      0   0  1
│ +  [15] .debug_aranges    PROGBITS        0000000000000000 048a45 0004a0 00      0   0  1
│ +  [16] .debug_line       PROGBITS        0000000000000000 048ee5 006a1e 00      0   0  1
│ +  [17] .debug_str        PROGBITS        0000000000000000 04f903 004950 01  MS  0   0  1
│ +  [18] .debug_line_str   PROGBITS        0000000000000000 054253 000596 01  MS  0   0  1
│ +  [19] .debug_rnglists   PROGBITS        0000000000000000 0547e9 0000ed 00      0   0  1
│ +  [20] .symtab           SYMTAB          0000000000000000 0548d8 0035e8 18     21 556  8
│ +  [21] .strtab           STRTAB          0000000000000000 057ec0 0010d4 00      0   0  1
│ +  [22] .shstrtab         STRTAB          0000000000000000 058f94 0000e4 00      0   0  1
│  Key to Flags:
[...]
```
(offsets are different, so there's a more output about each section)

File: src/boot/efi/meson.build [hunk 1]
```diff
@@ -304,6 +304,14 @@ if efi_arch[1] == 'arm'
         efi_ldflags += ['-Wl,--no-wchar-size-warning']
 endif
 
+if efi_arch[1] == 'aarch64'
+        # Newer binutils use a 64k alignemnt for aarch64 by default.
```
Comment (bluca) [inline src/boot/efi/meson.build]: s/alignemnt/alignment/

**Comment by medhefgo** [comment]
The differences seem to be just the in-file offsets for the ELF file. And we only care about the resulting *PE* files. And again: We disable this warning because it should have no relevance. So I want to know if there is an actual *runtime* issue with the final *EFI* binary or if someone is just re-enabling this warning for whatever reason (or if the warning detection code is failing).

**Comment by keszybz** [comment]
I checked, and indeed the warning only appears if our option to disable the warning is removed. So there is nothing to fix. Adding `max-page-size=4k` doesn't do anything for the warning.

Nevertheless, I think the patch is still marginally useful because it reduces the size of the produced file. (It's probably mostly zeros, but that still takes up space on disk or in memory.)  I repushed the patch with an adjusted justification.

**Comment by medhefgo** [comment]
But that size doesn't affect the PE binary. Who cares about some build artifact having a couple more bytes?

**Comment by keszybz** [comment]
We distribute the .elf file. It's even included in distro packages. I thought it's used for debugging of something, but I don't really know.

**Comment by bluca** [comment]
Yeah we ship the elf too in debian, just because it is installed by default by meson, not sure how/where it is used

**Comment by keszybz** [comment]
OK, let's close this then.

**Comment by maiomai** [comment]
I met `[EFI Stub ERROR kernel image not aligned on 64K boundary]` on a mkosi generated UKI. I wonder if we should set `-Wl,z,--max-page-size=65536` explicitly on aarch64 branch as I find the section alignment of systemd-stub delivered by Debian package systemd-boot-efi is still 4k. It seems the binutils does not do a 64k alignment by default.

**Pull request closed without merging**